configure.ac

```diff
@@ -247,6 +253,26 @@ if test "x$enable_debug" = xyes; then
     fi
 fi
 
+# Checking for -fsanitize flags works differently from other compiler flags.
+#
+# With GCC there's an implicit linking dependency on libasan, libtsan, etc. when
+# using the related -fsanitize flags, and those libraries are not included with
+# GCC on most Linux distros. The -fsanitize compiler options cause
+# GCC to invoke the linker with the required library for that flag (e.g. libasan
+# for -fsanitize=address). This means that AX_CHECK_COMPILE_FLAG does not work
+# properly with GCC, as the AX_CHECK_COMPILE_FLAG check here would pass if GCC
+# supports the flag, but in fact linking would fail.
+#
+# Clang works differently: it bundles these libraries, and statically links them
+# when using -fsanitize. For this reason, using AC_CHECK_LIB is not a good idea,
```

Comment:
I'm not sure what distinction you're trying to make here. As far as I can tell, clang/gcc handle this the same way. Of course, how distros configure and distribute them is a different story.

The following works fine for me with a gcc with busted libtsan:
```bash
AX_CHECK_LINK_FLAG([[-fsanitize=undefined]],[[SANITIZER_CXXFLAGS="${SANITIZER_CXXFLAGS} -fsanitize=undefined"; SANITIZER_LDFLAGS="${SANITIZER_LDFLAGS} -fsanitize=undefined"]],
[AC_MSG_ERROR([compiler does not support -fsanitize=undefined])], [[$CXXFLAG_WERROR]])

AX_CHECK_LINK_FLAG([[-fsanitize=thread]],[[SANITIZER_CXXFLAGS="${SANITIZER_CXXFLAGS} -fsanitize=thread"; SANITIZER_LDFLAGS="${SANITIZER_LDFLAGS} -fsanitize=thread"]],
[AC_MSG_ERROR([compiler does not support -fsanitize=thread])], [[$CXXFLAG_WERROR]])
```
The second check (correctly) fails due to a linking error with libtsan:
> configure:18545: checking whether the linker accepts -fsanitize=undefined
configure:18564: g++ -m64 -std=c++11 -o conftest -pipe -O2  -I/home/cory/dev/bitcoin2/depends/x86_64-pc-linux-gnu/share/../include/  -L/home/cory/dev/bitcoin2/depends/x86_64-pc-linux-gnu/share/../lib  -Werror -fsanitize=undefined conftest.cpp  >&5
configure:18564: $? = 0
configure:18574: result: yes
configure:18583: checking whether the linker accepts -fsanitize=thread
configure:18602: g++ -m64 -std=c++11 -o conftest -pipe -O2  -I/home/cory/dev/bitcoin2/depends/x86_64-pc-linux-gnu/share/../include/  -L/home/cory/dev/bitcoin2/depends/x86_64-pc-linux-gnu/share/../lib  -Werror -fsanitize=thread conftest.cpp  >&5
/tmp/ramfs/tmp/ccFEArS2.o: In function `_GLOBAL__sub_I_00099_0_main':
conftest.cpp:(.text.startup+0x11): undefined reference to `__tsan_init'
collect2: error: ld returned 1 exit status

I don't understand why you find AX_CHECK_LINK_FLAG insufficient, as it simply does a combined compile/link through the compiler, it doesn't invoke the linker directly.

Reply:
I'll try retesting this.